### PR TITLE
Fixed #25546 -- Prevented duplicate queries with nested prefetch_related().

### PR DESCRIPTION
**Pull request by francoisfreitag**
https://code.djangoproject.com/ticket/25546

@charettes: We already talked about this problem on IRC, here is a suggestion to fix it.

### CODE REVIEW

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by charettes** [comment]
@francoisfreitag could add another test method using `to_attr` like the reporter initially did?

File: tests/prefetch_related/tests.py [hunk 1]
```diff
@@ -1241,3 +1241,41 @@ def test_bug(self):
...
+                        Prefetch(
+                            'addresses',
+                            AuthorAddress.objects.filter(
+                                address="Happy place",
+                            ),
+                        )
+                    ),
+                ),
+            )
+            book1, book2 = list(books)
```
Comment (charettes) [inline tests/prefetch_related/tests.py]: Assert that accessing prefetched relationships doesn't issue queries (`self.assertNumQueries(0)`).

File: tests/prefetch_related/tests.py [hunk 2]
```diff
@@ -1241,3 +1241,41 @@ def test_bug(self):
         prefetcher = get_prefetcher(self.rooms[0], 'house')[0]
         queryset = prefetcher.get_prefetch_queryset(list(Room.objects.all()))[0]
         self.assertNotIn(' JOIN ', force_text(queryset.query))
+
+
+class Ticket25546Tests(TestCase):
+
+    def setUp(self):
```
Comment (charettes) [inline tests/prefetch_related/tests.py]: Use `setUpTestData` now that two models will be used by two methods.

**Comment by timgraham** [comment]
Could you add a docstring to the test class that summarizes the ticket so a reader doesn't have to look it up to understand what's being tested? You can also squash commits and use the commit message from the pull request title, thanks.

**Comment by timgraham** [comment]
merged in bdbe50a491ca41e7d4ebace47bfe8abe50a58211, thanks!

**Pull request closed without merging**